### PR TITLE
Improve documentation of keyname.c with suggestions from API review

### PR DESCRIPTION
**Pull request by lawli3t**
## Basics

These points need to be fulfilled for every PR:

- [ ] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [ ] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - [ ] add a line in `doc/news/_preparation_next_release.md`
  - [ ] reformat the code with `scripts/dev/reformat-all`
  - [ ] make all unit tests pass
  - [ ] fix all memleaks
- [ ] The PR is rebased with current master.

If you have any troubles fulfilling these criteria, please write
about the trouble as comment in the PR. We will help you.
But we cannot accept PRs that do not fulfill the basics.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned every code not directly written by me in [THIRD-PARTY-LICENSES](doc/THIRD-PARTY-LICENSES)

## Review

Reviewers will usually check the following:

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

If you are already Elektra developer:

- Add the "work in progress" label if you do not want the PR to be reviewed yet.
- Add the "ready to merge" label **if the basics are fulfilled** and you also
  say that everything is ready to be merged.

### CODE REVIEW

File: src/libs/elektra/keyname.c [hunk 1]
```diff
@@ -1229,21 +1276,28 @@ const char * keyBaseName (const Key * key)
 
 
 /**
- * Calculates number of bytes needed to store basename of @p key.
+ * Calculates number of bytes needed to store basename of @p key. (including
+ * NULL terminator)
```
Comment (mpranj) [inline src/libs/elektra/keyname.c:1280]:
```suggestion
 * Calculates number of bytes needed to store basename of @p key (including NULL terminator).
```

File: src/libs/elektra/keyname.c [hunk 2]
```diff
@@ -1255,23 +1309,32 @@ ssize_t keyGetBaseNameSize (const Key * key)
 
 
 /**
- * Calculate the basename of a key name and put it in @p returned finalizing
- * the string with NULL.
+ * Calculate the basename of a Key's name and copy it to @p returned
```
Comment (mpranj) [inline src/libs/elektra/keyname.c:1312]: I think this is much simpler:
```suggestion
 * Copy the Key's basename to @p returned
```
Any details could be in the long description below.

File: src/libs/elektra/keyname.c [hunk 3]
```diff
@@ -1528,17 +1591,20 @@ static size_t keyAddBaseNameInternal (Key * key, const char * baseName)
...
- * @param key the key object to work with
- * @param baseName the string to append to the name
- * @return the size in bytes of the new key name including the ending NULL
- * @retval -1 if the key had no name
+ * @return the size in bytes of the Key's new name including the NULL terminator
+ * @retval -1 if the Key has no name
  * @retval -1 on NULL pointers
- * @retval -1 if key was inserted to a keyset before
- * @retval -1 on allocation errors
+ * @retval -1 if Key was inserted to a KeySet before
```
Comment (mpranj) [inline src/libs/elektra/keyname.c:1600]: Maybe:
```suggestion
 * @retval -1 if Key was inserted into KeySet before
```

File: src/libs/elektra/keyname.c [hunk 4]
```diff
@@ -1552,17 +1618,17 @@ ssize_t keyAddBaseName (Key * key, const char * baseName)
 }
 
 /**
- * Sets @c baseName as the new basename for @c key.
+ * Sets @p baseName as the new basename for @p key.
  *
- * Only the baseName will be affected and no other part of the key.
+ * Only the basename will be affected and no other part of the Key.
```
Comment (mpranj) [inline src/libs/elektra/keyname.c:1623]: Sounds weird to me, here are two other variants.
Simple variant:
```suggestion
 * Only the basename of the Key will be affected.
```

Comment (mpranj) [reply src/libs/elektra/keyname.c:1623]: Second variant
```suggestion
 * Only the basename, and no other part of the Key, will be affected.
```

File: src/libs/elektra/keyname.c [hunk 5]
```diff
@@ -1580,23 +1646,24 @@ ssize_t keyAddBaseName (Key * key, const char * baseName)
...
+ * @param baseName the new basename for the Key
  *
- *
- * @param key the key object to work with
- * @param baseName the string used to overwrite the basename of the key
  * @return the size in bytes of the new key name
- * @retval -1 on NULL pointers in key
- * @retval -1 if key was inserted to a keyset before
+ * @retval -1 on NULL pointers
+ * @retval -1 if Key was inserted to a KeySet before
```
Comment (mpranj) [inline src/libs/elektra/keyname.c:1657]: Maybe
```suggestion
 * @retval -1 if Key was inserted into KeySet before
```

File: src/libs/elektra/keyname.c [hunk 6]
```diff
@@ -1631,25 +1698,28 @@ ssize_t keySetBaseName (Key * key, const char * baseName)
 }
 
 /**
- * For currently valid namespaces see #elektraNamespace.
+ * Returns the #elektraNamespace for a Key.
```
Comment (mpranj) [inline src/libs/elektra/keyname.c:1701]: General note: we should update the docu of `#elektraNamespace`, maybe link to Key Name docu (`doc/KEYNAMES.md`). The Key Name docu explains namespaces with more detail.

**Review by mpranj — APPROVED** [review]
Great work! Thank you so much!

I added some feedback and optional improvements. Select or adapt what you agree with.

**Comment by mpranj** [comment]
Needs rebase & has open conversations from review.

File: src/libs/elektra/keyname.c [hunk 7]
```diff
@@ -325,30 +341,40 @@ ssize_t keyGetUnescapedNameSize (const Key * key)
 
 
 /**
- * Get abbreviated key name (without owner name).
+ * Get abbreviated Key name (excluding owner).
  *
  * When there is not enough space to write the name,
  * nothing will be written and -1 will be returned.
  *
  * maxSize is limited to SSIZE_MAX. When this value
```
[the remark this answers is not in the document]
Comment (lawli3t) [reply src/libs/elektra/keyname.c:349]: `@p maxSize` seems good, sadly it is really hard to catch all mentions of parameters all the time - good catch!

`@p SSIZE_MAX` seems to be a bit awkward, `@p` is used for function parameters if I am not mistaken.

File: src/libs/elektra/keyname.c [hunk 8]
```diff
@@ -426,30 +462,34 @@ ssize_t keyGetUnescapedName (const Key * key, char * returnedName, size_t maxSiz
...
+ * (name will be empty afterwards)
  * @retval -1 if newName is invalid (name will be empty afterwards)
- * @retval -1 if key was inserted to a keyset before
- * @param key the key object to work with
- * @param newName the new key name
- * @see keyNew()
- * @see keyGetName(), keyName()
- * @see keySetBaseName(), keyAddBaseName() to manipulate a name
+ * @retval -1 if Key was inserted to a KeySet before
+ * @retval -1 if Key is read-only
```
[the remark this answers is not in the document]
Comment (lawli3t) [reply src/libs/elektra/keyname.c:483]: Yes, its always hard to strike a balance between the amount of cases in `@retval` :sweat_smile: 

I have tried to consolidate it, but for this module it was really hard for me. I hope the new version I am about to push alleviates this somewhat (for this function at least).

File: src/libs/elektra/keyname.c [hunk 9]
```diff
@@ -377,21 +403,31 @@ ssize_t keyGetName (const Key * key, char * returnedName, size_t maxSize)
 }
 
 /**
- * Copies the unescaped name of a key into provided buffer.
- * We will only copy the full name, if the buffer is to small an error code is returned.
+ * Copies the unescaped name of a Key into @p returnedName
```
[the remark this answers is not in the document]
Comment (lawli3t) [reply src/libs/elektra/keyname.c]: changed with https://github.com/ElektraInitiative/libelektra/pull/3795/commits/cea2c469decb39e99b3d71197f27ed40dede38d9

File: src/libs/elektra/keyname.c [hunk 10]
```diff
@@ -426,30 +462,34 @@ ssize_t keyGetUnescapedName (const Key * key, char * returnedName, size_t maxSiz
...
  *
  * .., . and / will be handled as in filesystem paths. A valid name will be build
  * out of the (valid) name what you pass, e.g. user:///sw/../sw//././MyApp -> user:/sw/MyApp
  *
  * On invalid names, NULL or "" the name will be "" afterwards.
  *
- *
- * @retval size in bytes of this new key name including ending NULL
- * @retval 0 if newName is an empty string or a NULL pointer (name will be empty afterwards)
+ * @return size in bytes of the new Key name, including ending NULL
```
[the remark this answers is not in the document]
Comment (lawli3t) [reply src/libs/elektra/keyname.c]: fixed in https://github.com/ElektraInitiative/libelektra/pull/3795/commits/cea2c469decb39e99b3d71197f27ed40dede38d9

File: src/libs/elektra/keyname.c (hunk 1 above)
Comment (lawli3t) [reply src/libs/elektra/keyname.c]: fixed in https://github.com/ElektraInitiative/libelektra/pull/3795/commits/cea2c469decb39e99b3d71197f27ed40dede38d9

File: src/libs/elektra/keyname.c (hunk 2 above)
Comment (lawli3t) [reply src/libs/elektra/keyname.c]: fixed in https://github.com/ElektraInitiative/libelektra/pull/3795/commits/cea2c469decb39e99b3d71197f27ed40dede38d9

File: src/libs/elektra/keyname.c (hunk 3 above)
Comment (lawli3t) [reply src/libs/elektra/keyname.c]: fixed in https://github.com/ElektraInitiative/libelektra/pull/3795/commits/cea2c469decb39e99b3d71197f27ed40dede38d9

File: src/libs/elektra/keyname.c (hunk 4 above)
Comment (lawli3t) [reply src/libs/elektra/keyname.c]: First variant seems good to me - Occam's razor and stuff :wink:

File: src/libs/elektra/keyname.c (hunk 5 above)
Comment (lawli3t) [reply src/libs/elektra/keyname.c]: fixed in https://github.com/ElektraInitiative/libelektra/pull/3795/commits/cea2c469decb39e99b3d71197f27ed40dede38d9

File: src/libs/elektra/keyname.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (mpranj) [reply src/libs/elektra/keyname.c]: Afaik `@p` is used to set a typewriter font, not to mark function parameters.

**Comment by mpranj** [comment]
Thank you!

**Pull request merged**